File: code/packages/lua/b-tree/tests/test_b_tree.lua
-- tests/test_b_tree.lua — Comprehensive tests for the B-Tree (DT11)
-- =================================================================
--
-- Uses the Busted test framework.  Run from the tests/ directory:
--   LUA_PATH="../src/?.lua;../src/?/init.lua;;" busted . --verbose

package.path = "../src/?.lua;../src/?/init.lua;" .. package.path

local BTree = require("coding_adventures.b_tree")

-- ============================================================================
-- Helpers
-- ============================================================================

--- Assert that tree.is_valid() returns true.
local function check(tree, label)
    assert.is_true(tree:is_valid(), "is_valid: " .. (label or ""))
end

--- Assert that inorder returns a sorted array of keys.
local function assert_sorted(pairs_list, label)
    for i = 2, #pairs_list do
        assert.is_true(
            pairs_list[i][1] > pairs_list[i - 1][1],
            "out of order at index " .. i .. ": " .. (label or "")
        )
    end
end

-- ============================================================================
-- 1. Empty tree
-- ============================================================================
describe("empty tree", function()
    it("has size 0", function()
        local t = BTree.new()
        assert.equals(0, t:size())
    end)
    it("has height 0", function()
        local t = BTree.new()
        assert.equals(0, t:height())
    end)
    it("search returns nil", function()
        local t = BTree.new()
        assert.is_nil(t:search("x"))
    end)
    it("min_key returns nil", function()
        local t = BTree.new()
        assert.is_nil(t:min_key())
    end)
    it("max_key returns nil", function()
        local t = BTree.new()
        assert.is_nil(t:max_key())
    end)
    it("inorder returns empty list", function()
        local t = BTree.new()
        assert.same({}, t:inorder())
    end)
    it("delete returns false", function()
        local t = BTree.new()
        assert.is_false(t:delete(42))
    end)
    it("is_valid", function()
        local t = BTree.new()
        check(t, "empty")
    end)
end)

-- ============================================================================
-- 2. Basic insert / search (t=2)
-- ============================================================================
describe("basic insert and search t=2", function()
    local t
    before_each(function()
        t = BTree.new({ t = 2 })
        local pairs = {{10,"ten"},{20,"twenty"},{5,"five"},{15,"fifteen"},{25,"twenty-five"}}
        for _, p in ipairs(pairs) do t:insert(p[1], p[2]) end
    end)
    it("size = 5", function()
        assert.equals(5, t:size())
    end)
    it("finds each inserted key", function()
        assert.equals("ten",         t:search(10))
        assert.equals("twenty",      t:search(20))
        assert.equals("five",        t:search(5))
        assert.equals("fifteen",     t:search(15))
        assert.equals("twenty-five", t:search(25))
    end)
    it("absent key returns nil", function()
        assert.is_nil(t:search(99))
    end)
    it("min_key = 5", function()
        assert.equals(5, t:min_key())
    end)
    it("max_key = 25", function()
        assert.equals(25, t:max_key())
    end)
    it("is_valid", function()
        check(t, "after inserts")
    end)
end)

-- ============================================================================
-- 3. Upsert
-- ============================================================================
describe("upsert", function()
    it("updates value for existing key", function()
        local t = BTree.new()
        t:insert(1, "one")
        t:insert(1, "ONE")
        assert.equals(1, t:size())
        assert.equals("ONE", t:search(1))
        check(t, "upsert")
    end)
end)

-- ============================================================================
-- 4. Inorder traversal is sorted
-- ============================================================================
describe("inorder sorted t=3", function()
    it("returns pairs in ascending order", function()
        local t = BTree.new({ t = 3 })
        for _, k in ipairs({50, 10, 80, 30, 60, 5, 25, 55, 75, 90}) do
            t:insert(k, k * 2)
        end
        local pairs = t:inorder()
        assert_sorted(pairs, "inorder")
        for _, p in ipairs(pairs) do
            assert.equals(p[1] * 2, p[2])
        end
        check(t, "inorder")
    end)
end)

-- ============================================================================
-- 5. Range query
-- ============================================================================
describe("range_query", function()
    local t
    before_each(function()
        t = BTree.new()
        for i = 1, 20 do t:insert(i, "v" .. i) end
    end)
    it("range 5..10", function()
        local r = t:range_query(5, 10)
        assert.equals(6, #r)
        for i, p in ipairs(r) do
            assert.equals(4 + i, p[1])
        end
    end)
    it("empty range returns empty list", function()
        assert.same({}, t:range_query(100, 200))
    end)
    it("single-element range", function()
        local r = t:range_query(7, 7)
        assert.equals(1, #r)
        assert.equals(7, r[1][1])
    end)
    it("is_valid", function()
        check(t, "range_query")
    end)
end)

-- ============================================================================
-- 6. Delete — leaf (Case A)
-- ============================================================================
describe("delete leaf", function()
    it("removes the key and returns true", function()
        local t = BTree.new()
        for _, k in ipairs({10, 20, 5, 15, 25}) do t:insert(k, k) end
        assert.is_true(t:delete(5))
        assert.equals(4, t:size())
        assert.is_nil(t:search(5))
        assert.is_false(t:delete(99))
        check(t, "delete leaf")
    end)
end)

-- ============================================================================
-- 7. Delete — internal node (Cases B1/B2/B3)
-- ============================================================================
describe("delete internal node", function()
    it("removes internal key correctly", function()
        local t = BTree.new({ t = 2 })
        for k = 1, 15 do t:insert(k, k) end
        check(t, "before internal delete")
        assert.is_true(t:delete(8))
        assert.is_nil(t:search(8))
        assert.equals(14, t:size())
        check(t, "after internal delete")
        t:delete(4); t:delete(12)
        check(t, "after more deletes")
    end)
end)

-- ============================================================================
-- 8. Delete all cases (1..30 then specific order)
-- ============================================================================
describe("delete all cases", function()
    it("handles all CLRS deletion sub-cases", function()
        local t = BTree.new({ t = 2 })
        for k = 1, 30 do t:insert(k, k) end
        check(t, "after 30 inserts")

        local order = {15, 1, 30, 10, 20, 5, 25, 8, 22, 3, 17, 28}
        for _, k in ipairs(order) do
            assert.is_true(t:delete(k), "delete " .. k)
            assert.is_nil(t:search(k),  "key " .. k .. " gone")
            check(t, "after delete " .. k)
        end
        assert.equals(30 - 12, t:size())
    end)
end)

-- ============================================================================
-- 9. Delete until empty
-- ============================================================================
describe("delete until empty", function()
    it("maintains validity at every step", function()
        local t = BTree.new()
        for k = 1, 10 do t:insert(k, k) end
        for k = 1, 10 do
            t:delete(k)
            check(t, "after delete " .. k)
        end
        assert.equals(0, t:size())
        assert.is_nil(t:min_key())
        assert.is_nil(t:max_key())
    end)
end)

-- ============================================================================
-- 10. Large-scale — 500 keys, t=2
-- ============================================================================
describe("large scale t=2 (500 keys)", function()
    it("insert, search, and delete are correct", function()
        local t = BTree.new({ t = 2 })
        -- Fisher-Yates shuffle.
        local keys = {}
        for i = 1, 500 do keys[i] = i end
        for i = 500, 2, -1 do
            local j = math.random(i)
            keys[i], keys[j] = keys[j], keys[i]
        end
        for _, k in ipairs(keys) do t:insert(k, k * 3) end
        assert.equals(500, t:size())
        check(t, "500 inserts")
        assert.equals(1,   t:min_key())
        assert.equals(500, t:max_key())
        for k = 1, 500 do
            assert.equals(k * 3, t:search(k))
        end
        -- Delete odd keys.
        for k = 1, 500, 2 do t:delete(k) end
        assert.equals(250, t:size())
        check(t, "after deleting 250 keys")
    end)
end)

-- ============================================================================
-- 11. Large-scale — 600 keys, t=3
-- ============================================================================
describe("large scale t=3 (600 keys)", function()
    it("inorder is 1..600 and deletion is correct", function()
        local t = BTree.new({ t = 3 })
        for k = 1, 600 do t:insert(k, k) end
        assert.equals(600, t:size())
        check(t, "600 inserts")

        local pairs = t:inorder()
        assert.equals(600, #pairs)
        for i, p in ipairs(pairs) do
            assert.equals(i, p[1])
        end

        -- Delete even keys.
        for k = 2, 600, 2 do t:delete(k) end
        assert.equals(300, t:size())
        check(t, "after halving")
    end)
end)

-- ============================================================================
-- 12. Large-scale — 750 keys, t=5
-- ============================================================================
describe("large scale t=5 (750 keys)", function()
    it("range query is correct", function()
        local t = BTree.new({ t = 5 })
        for k = 1, 750 do t:insert(k, "key" .. k) end
        assert.equals(750, t:size())
        check(t, "750 inserts")

        local r = t:range_query(100, 200)
        assert.equals(101, #r)
        assert.equals(100, r[1][1])
        assert.equals(200, r[#r][1])
    end)
end)

-- ============================================================================
-- 13. Min / max after deletions
-- ============================================================================
describe("min max after deletions", function()
    it("updates correctly", function()
        local t = BTree.new()
        for _, k in ipairs({7, 3, 11, 1, 5, 9, 13}) do t:insert(k, k) end
        assert.equals(1,  t:min_key())
        assert.equals(13, t:max_key())
        t:delete(1)
        assert.equals(3,  t:min_key())
        t:delete(13)
        assert.equals(11, t:max_key())
        check(t, "min max")
    end)
end)

-- ============================================================================
-- 14. Reverse-order insert
-- ============================================================================
describe("reverse order insert t=3", function()
    it("produces sorted inorder traversal", function()
        local t = BTree.new({ t = 3 })
        for k = 300, 1, -1 do t:insert(k, k) end
        assert.equals(300, t:size())
        check(t, "reverse inserts")
        local pairs = t:inorder()
        assert.equals(300, #pairs)
        for i, p in ipairs(pairs) do
            assert.equals(i, p[1])
        end
    end)
end)

-- ============================================================================
-- 15. Insert → delete all → re-insert
-- ============================================================================
describe("insert delete reinsert", function()
    it("recovers correctly after full deletion", function()
        local t = BTree.new()
        for k = 1, 50 do t:insert(k, k) end
        for k = 1, 50 do t:delete(k) end
        assert.equals(0, t:size())
        check(t, "after all deletions")
        for k = 51, 100 do t:insert(k, k) end
        assert.equals(50,  t:size())
        check(t, "after reinsert")
        assert.equals(51,  t:min_key())
        assert.equals(100, t:max_key())
    end)
end)

File: code/packages/python/b-tree/tests/test_b_tree.py
"""Comprehensive tests for the BTree implementation.

Test strategy:
  - Every operation is followed by tree.is_valid() to ensure structural integrity
  - Delete cases 1 (leaf), 2a/2b/2c (internal), and 3a/3b (pre-fill) are
    explicitly covered
  - Tests run with t=2, t=3, and t=5 to cover different node capacities
  - A 1000-key stress test verifies large-scale correctness
"""

import pytest

from b_tree import BTree, BTreeNode


# ---------------------------------------------------------------------------
# Construction
# ---------------------------------------------------------------------------

class TestConstruction:
    def test_default_degree(self) -> None:
        tree = BTree()
        assert tree._t == 2

    def test_custom_degree(self) -> None:
        tree = BTree(t=5)
        assert tree._t == 5

    def test_invalid_degree_raises(self) -> None:
        with pytest.raises(ValueError):
            BTree(t=1)
        with pytest.raises(ValueError):
            BTree(t=0)

    def test_empty_tree_len(self) -> None:
        assert len(BTree()) == 0

    def test_empty_tree_bool(self) -> None:
        assert not BTree()

    def test_empty_tree_valid(self) -> None:
        assert BTree().is_valid()

    def test_empty_tree_search(self) -> None:
        assert BTree().search(42) is None

    def test_empty_min_raises(self) -> None:
        with pytest.raises(ValueError):
            BTree().min_key()

    def test_empty_max_raises(self) -> None:
        with pytest.raises(ValueError):
            BTree().max_key()


# ---------------------------------------------------------------------------
# Basic insert / search
# ---------------------------------------------------------------------------

class TestInsertSearch:
    def test_single_insert(self) -> None:
        tree = BTree(t=2)
        tree.insert(10, "ten")
        assert tree.search(10) == "ten"
        assert len(tree) == 1
        assert tree.is_valid()

    def test_bool_nonempty(self) -> None:
        tree = BTree()
        tree.insert(1, "a")
        assert bool(tree)

    def test_update_existing_key(self) -> None:
        tree = BTree(t=2)
        tree.insert(5, "five")
        tree.insert(5, "FIVE")
        assert tree.search(5) == "FIVE"
        assert len(tree) == 1
        assert tree.is_valid()

    def test_search_missing_key(self) -> None:
        tree = BTree(t=2)
        tree.insert(1, "a")
        assert tree.search(99) is None

    def test_contains_operator(self) -> None:
        tree = BTree(t=2)
        tree.insert(7, "seven")
        assert 7 in tree
        assert 8 not in tree

    def test_getitem(self) -> None:
        tree = BTree(t=2)
        tree.insert(3, "three")
        assert tree[3] == "three"

    def test_getitem_missing_raises(self) -> None:
        tree = BTree(t=2)
        with pytest.raises(KeyError):
            _ = tree[99]

    def test_setitem(self) -> None:
        tree = BTree(t=2)
        tree[42] = "answer"
        assert tree.search(42) == "answer"
        assert tree.is_valid()

    def test_insert_sequential(self) -> None:
        """Insert 1..20 in order and verify all are findable."""
        tree = BTree(t=2)
        for i in range(1, 21):
            tree.insert(i, i * 10)
        for i in range(1, 21):
            assert tree.search(i) == i * 10
        assert len(tree) == 20
        assert tree.is_valid()

    def test_insert_reverse(self) -> None:
        """Insert in reverse order — forces splits from the right."""
        tree = BTree(t=2)
        for i in range(20, 0, -1):
            tree.insert(i, str(i))
        assert len(tree) == 20
        assert tree.is_valid()

    def test_insert_random_order(self) -> None:
        import random
        keys = list(range(50))
        random.shuffle(keys)
        tree = BTree(t=3)
        for k in keys:
            tree.insert(k, k)
        assert len(tree) == 50
        assert tree.is_valid()


# ---------------------------------------------------------------------------
# min / max
# ---------------------------------------------------------------------------

class TestMinMax:
    def test_single_element(self) -> None:
        tree = BTree(t=2)
        tree.insert(7, "x")
        assert tree.min_key() == 7
        assert tree.max_key() == 7

    def test_multiple_elements(self) -> None:
        tree = BTree(t=2)
        for k in [5, 2, 8, 1, 9]:
            tree.insert(k, k)
        assert tree.min_key() == 1
        assert tree.max_key() == 9

    def test_after_delete_min(self) -> None:
        tree = BTree(t=2)
        for k in [3, 1, 2]:
            tree.insert(k, k)
        tree.delete(1)
        assert tree.min_key() == 2
        assert tree.is_valid()

    def test_after_delete_max(self) -> None:
        tree = BTree(t=2)
        for k in [3, 1, 2]:
            tree.insert(k, k)
        tree.delete(3)
        assert tree.max_key() == 2
        assert tree.is_valid()


# ---------------------------------------------------------------------------
# Inorder / range_query
# ---------------------------------------------------------------------------

class TestInorderRange:
    def test_inorder_empty(self) -> None:
        assert list(BTree().inorder()) == []

    def test_inorder_sorted(self) -> None:
        tree = BTree(t=2)
        for k in [5, 3, 7, 1, 9]:
            tree.insert(k, k * 10)
        result = list(tree.inorder())
        assert result == [(1, 10), (3, 30), (5, 50), (7, 70), (9, 90)]

    def test_range_empty(self) -> None:
        tree = BTree(t=2)
        tree.insert(10, "a")
        assert tree.range_query(20, 30) == []

    def test_range_all(self) -> None:
        tree = BTree(t=2)
        for k in [1, 2, 3]:
            tree.insert(k, str(k))
        assert tree.range_query(0, 10) == [(1, "1"), (2, "2"), (3, "3")]

    def test_range_subset(self) -> None:
        tree = BTree(t=3)
        for k in [1, 3, 5, 7, 9]:
            tree.insert(k, k)
        result = tree.range_query(3, 7)
        assert result == [(3, 3), (5, 5), (7, 7)]

    def test_range_inclusive_boundaries(self) -> None:
        tree = BTree(t=2)
        for k in [1, 2, 3, 4, 5]:
            tree.insert(k, k)
        assert tree.range_query(2, 4) == [(2, 2), (3, 3), (4, 4)]


# ---------------------------------------------------------------------------
# Height
# ---------------------------------------------------------------------------

class TestHeight:
    def test_empty_height(self) -> None:
        assert BTree().height() == 0

    def test_single_node_height(self) -> None:
        tree = BTree(t=2)
        for i in range(1, 4):
            tree.insert(i, i)
        # With t=2, 3 keys fit in the root (max 3 keys = 2t-1)
        assert tree.height() == 0

    def test_height_grows_on_split(self) -> None:
        """With t=2 and 4 keys, the root must split → height 1."""
        tree = BTree(t=2)
        for i in range(1, 5):
            tree.insert(i, i)
        assert tree.height() >= 1
        assert tree.is_valid()

    def test_all_leaves_same_depth(self) -> None:
        """is_valid() enforces this — just exercise it with a large tree."""
        tree = BTree(t=3)
        for i in range(100):
            tree.insert(i, i)
        assert tree.is_valid()


# ---------------------------------------------------------------------------
# Delete — Case 1: delete from leaf
# ---------------------------------------------------------------------------

class TestDeleteLeaf:
    def setup_method(self) -> None:
        """Build a small tree for leaf-delete tests."""
        self.tree = BTree(t=2)
        for k in [10, 20, 5, 15, 25]:
            self.tree.insert(k, str(k))

    def test_delete_leaf_key(self) -> None:
        self.tree.delete(5)
        assert 5 not in self.tree
        assert len(self.tree) == 4
        assert self.tree.is_valid()

    def test_delete_missing_raises(self) -> None:
        with pytest.raises(KeyError):
            self.tree.delete(99)

    def test_delitem_operator(self) -> None:
        del self.tree[20]
        assert 20 not in self.tree
        assert self.tree.is_valid()

    def test_delete_all_leaves(self) -> None:
        """Delete every key; tree should become empty and still valid."""
        keys = [10, 20, 5, 15, 25]
        for k in keys:
            self.tree.delete(k)
        assert len(self.tree) == 0
        assert self.tree.is_valid()


# ---------------------------------------------------------------------------
# Delete — Case 2: delete from internal node
# ---------------------------------------------------------------------------

class TestDeleteInternal:
    def _make_tree(self, keys: list[int]) -> BTree:
        tree = BTree(t=2)
        for k in keys:
            tree.insert(k, str(k))
        return tree

    def test_case_2a_predecessor(self) -> None:
        """Delete an internal node key when left child has >= t keys."""
        # Build [1..7] with t=2 → root has 1 key (4), left child [1,2,3], right [5,6,7]
        tree = self._make_tree(list(range(1, 8)))
        # Deleting root key should trigger Case 2a or 2b
        root_key = tree._root.keys[0]
        tree.delete(root_key)
        assert root_key not in tree
        assert tree.is_valid()

    def test_case_2b_successor(self) -> None:
        """Force Case 2b by ensuring left child is thin (t-1 keys)."""
        tree = BTree(t=2)
        for k in [2, 1, 3, 4, 5]:
            tree.insert(k, str(k))
        # Find an internal key and delete it; test that the tree stays valid
        # regardless of which case fires
        tree.delete(2)
        assert 2 not in tree
        assert tree.is_valid()

    def test_case_2c_merge(self) -> None:
        """Force Case 2c: both children thin → merge."""
        tree = BTree(t=2)
        for k in range(1, 8):
            tree.insert(k, str(k))
        # Delete keys until internal nodes are thin enough to force Case 2c
        # Then delete the internal key itself
        tree.delete(6)
        tree.delete(7)
        # Now right child of root[0] should be thin; deleting root[0] → merge
        tree.delete(tree._root.keys[0])
        assert tree.is_valid()

    def test_delete_all_internal(self) -> None:
        """Large tree: delete all keys one by one."""
        tree = BTree(t=3)
        keys = list(range(30))
        for k in keys:
            tree.insert(k, k)
        import random
        random.shuffle(keys)
        for k in keys:
            tree.delete(k)
            assert tree.is_valid()
        assert len(tree) == 0


# ---------------------------------------------------------------------------
# Delete — Case 3: pre-fill before descent
# ---------------------------------------------------------------------------

class TestDeletePreFill:
    def test_case_3a_rotate_from_left(self) -> None:
        """The child being descended into is thin; borrow from left sibling."""
        tree = BTree(t=2)
        for k in range(1, 10):
            tree.insert(k, str(k))
        # After building, delete keys that force pre-fill rotations
        tree.delete(1)
        tree.delete(2)
        assert tree.is_valid()

    def test_case_3a_rotate_from_right(self) -> None:
        tree = BTree(t=2)
        for k in range(1, 10):
            tree.insert(k, str(k))
        tree.delete(9)
        tree.delete(8)
        assert tree.is_valid()

    def test_case_3b_merge_siblings(self) -> None:
        """Both siblings are thin → merge triggers."""
        tree = BTree(t=2)
        for k in [1, 2, 3, 4, 5, 6, 7]:
            tree.insert(k, str(k))
        # Build a state where siblings have minimum keys, then delete
        tree.delete(3)
        tree.delete(4)
        tree.delete(5)
        assert tree.is_valid()

    def test_root_shrinks_height(self) -> None:
        """After merging all root's children, root becomes empty and is replaced."""
        tree = BTree(t=2)
        for k in range(1, 5):
            tree.insert(k, str(k))
        # All deletes
        for k in range(1, 5):
            tree.delete(k)
        assert len(tree) == 0
        assert tree.is_valid()


# ---------------------------------------------------------------------------
# Different degrees
# ---------------------------------------------------------------------------

class TestDegrees:
    def test_t2(self) -> None:
        tree = BTree(t=2)
        for k in range(50):
            tree.insert(k, k)
        assert tree.is_valid()
        assert list(tree.inorder()) == [(k, k) for k in range(50)]

    def test_t3(self) -> None:
        tree = BTree(t=3)
        for k in range(100):
            tree.insert(k, k)
        assert tree.is_valid()
        for k in range(100):
            assert tree.search(k) == k

    def test_t5(self) -> None:
        tree = BTree(t=5)
        for k in range(200):
            tree.insert(k, k)
        assert tree.is_valid()
        for k in range(0, 200, 7):
            tree.delete(k)
        assert tree.is_valid()


# ---------------------------------------------------------------------------
# Large-scale stress test
# ---------------------------------------------------------------------------

class TestLargeScale:
    def test_1000_keys_insert_search(self) -> None:
        tree = BTree(t=3)
        for i in range(1000):
            tree.insert(i, i * 2)
        assert len(tree) == 1000
        assert tree.is_valid()
        for i in range(1000):
            assert tree.search(i) == i * 2

    def test_1000_keys_delete_all(self) -> None:
        import random
        tree = BTree(t=3)
        keys = list(range(1000))
        for k in keys:
            tree.insert(k, k)
        random.shuffle(keys)
        for k in keys:
            tree.delete(k)
        assert len(tree) == 0
        assert tree.is_valid()

    def test_1000_keys_inorder_sorted(self) -> None:
        tree = BTree(t=4)
        import random
        keys = list(range(1000))
        random.shuffle(keys)
        for k in keys:
            tree.insert(k, k)
        result = [k for k, _ in tree.inorder()]
        assert result == sorted(keys)

    def test_1000_keys_range_query(self) -> None:
        tree = BTree(t=3)
        for k in range(1000):
            tree.insert(k, k)
        result = tree.range_query(200, 300)
        assert len(result) == 101
        assert result[0] == (200, 200)
        assert result[-1] == (300, 300)

    def test_interleaved_insert_delete(self) -> None:
        """Interleave inserts and deletes; tree must always remain valid."""
        tree = BTree(t=2)
        for i in range(500):
            tree.insert(i, i)
            if i % 3 == 0 and i > 0:
                tree.delete(i - 3)
            assert tree.is_valid()

    def test_duplicate_inserts(self) -> None:
        tree = BTree(t=3)
        for _ in range(5):
            for k in range(100):
                tree.insert(k, k * 10)
        assert len(tree) == 100  # duplicates update, don't add
        assert tree.is_valid()


# ---------------------------------------------------------------------------
# BTreeNode unit tests
# ---------------------------------------------------------------------------

class TestBTreeNode:
    def test_is_full(self) -> None:
        node = BTreeNode(keys=[1, 2, 3])
        assert node.is_full(t=2)
        node2 = BTreeNode(keys=[1, 2])
        assert not node2.is_full(t=2)

    def test_find_key_index_exact(self) -> None:
        node = BTreeNode(keys=[10, 20, 30])
        assert node.find_key_index(20) == 1

    def test_find_key_index_between(self) -> None:
        node = BTreeNode(keys=[10, 20, 30])
        assert node.find_key_index(15) == 1  # between 10 and 20

    def test_find_key_index_before_all(self) -> None:
        node = BTreeNode(keys=[10, 20, 30])
        assert node.find_key_index(5) == 0

    def test_find_key_index_after_all(self) -> None:
        node = BTreeNode(keys=[10, 20, 30])
        assert node.find_key_index(35) == 3


# ---------------------------------------------------------------------------
# Edge cases
# ---------------------------------------------------------------------------

class TestEdgeCases:
    def test_insert_none_value(self) -> None:
        """Values can be None (e.g., set-like usage)."""
        tree = BTree(t=2)
        tree.insert(1, None)
        assert 1 in tree
        # __getitem__ should return None for a key with None value
        assert tree[1] is None

    def test_delete_reduces_size(self) -> None:
        tree = BTree(t=2)
        for k in range(10):
            tree.insert(k, k)
        for k in range(10):
            tree.delete(k)
            assert len(tree) == 9 - k

    def test_repr(self) -> None:
        tree = BTree(t=2)
        r = repr(tree)
        assert "BTree" in r

    def test_range_single_element(self) -> None:
        tree = BTree(t=2)
        tree.insert(5, "five")
        assert tree.range_query(5, 5) == [(5, "five")]

    def test_range_no_results(self) -> None:
        tree = BTree(t=2)
        tree.insert(1, "one")
        assert tree.range_query(2, 10) == []

    def test_update_then_delete(self) -> None:
        tree = BTree(t=2)
        tree.insert(42, "old")
        tree.insert(42, "new")
        assert tree[42] == "new"
        tree.delete(42)
        assert 42 not in tree
        assert tree.is_valid()

    def test_string_keys(self) -> None:
        tree = BTree(t=2)
        tree.insert("banana", 1)
        tree.insert("apple", 2)
        tree.insert("cherry", 3)
        assert tree.min_key() == "apple"
        assert tree.max_key() == "cherry"
        assert tree.is_valid()

    def test_height_single_leaf(self) -> None:
        tree = BTree(t=2)
        tree.insert(1, 1)
        assert tree.height() == 0
